lib/swagger-catalog.js

```diff
@@ -25,57 +33,134 @@ function getServiceSummary(service) {
+          for (key in tempPaths) {
+            openApi.paths["/" + service.serviceName + "/" + tempVersion + key] = tempPaths[key];
           }
+        });
+      } else {
+        // if there are no swagger docs that were read use this placeholder setup
+        for (const service of plugin.dataServices || []) {
+          //FIXME templates with an asterisk are not supported by open api
+          //TODO we can actually somewhat inspect Express routers
+          openApi.paths[zLuxUrl.makeServiceSubURL(service) + '/*'] = {
+            summary: getServiceSummary(service),
+            get: {
+              responses: {
+                200: {
+                  description: "service call succeeded"
+                }
+              }
+            },
+            post: {
+              responses: {
+                200: {
+                  description: "service call succeeded"
+                }
+              }
+            },
+            put: {
+              responses: {
+                200: {
+                  description: "service call succeeded"
+                }
+              }
+            },
+            delete: {
+              responses: {
+                200: {
+                  description: "service call succeeded"
+                }
+              }
+            }
+          };
         }
-      },
-      delete: {
-        responses: {
-          200: {
-            description: "service call succeeded"
-          }
+      }
+      // pass everything back at once the plugin catalog with all services together
+      // swaggerdocs is the list of all the docs for each service
+      var allDocumentation = {
+        pluginCatalog: openApi,
+        serviceDocs: swaggerDocs
+      }
+      resolve(allDocumentation)
+    });
+  })
+}
+
+async function getSwaggerDocs(plugin, productCode, nodeContext) {
+  var allServiceDocs = [];
+  if (plugin.dataServices){
+    for (let i = 0; i < plugin.dataServices.length; i++) {
+      let service = plugin.dataServices[i]
+      if (service.swaggerdoc) {
+        let fileContent = await readSingleSwaggerFile(path.join(plugin.location, "doc/swagger"), service.swaggerdoc)
+          .catch((err) => {
+            installLog.warn('Invalid Swagger from file ' + service.swaggerdoc)
+            installLog.warn(err);
+          })
+        if (fileContent) {
+          fileContent = overwriteSwaggerFieldsForServer(fileContent, 
+            zLuxUrl.makePluginURL(productCode, plugin.identifier), nodeContext);
+          allServiceDocs.push({
+            "serviceName" : service.name,
+            "serviceDoc" : fileContent
+          })
         }
-      },
+      }
     }
   }
-  return openApi;
+  return allServiceDocs;
+}
+
+function readSingleSwaggerFile (dirName, fileName) {
+  // read one swagger file and validate the json that is returned
+  return new Promise ((resolve, reject) => {
+    let fileContent = fs.readFileSync(path.join(dirName, fileName), 'utf-8');
```

Comment:
Promise allows asynchrony, so readFileSync just slows everything down. Please do one or the other, don't mix them. Only do sync if there is some timing issue, and even then it is best to plan a fix.